### PR TITLE
Add a reference example for PSA_ALG_SP800_108_COUNTER_HMAC/CMAC

### PR DESCRIPTION
**Pull request by mswarowsky**
Adds a reference implementation for a counter mode KDF using the construction recommended by NIST SP 800-108, introduced in #123.

The example is using the PSA API's with mbedtls implementation.

### CODE REVIEW

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:3]: Please use an SPDX file copyright tag, either to replace or in addition to this comment. E.g.

```suggestion
 * SPDX-FileCopyrightText: Copyright (c) 2023 Nordic Semiconductor ASA
```

File: examples/crypto/SP800-108_counter_KDF/README.md
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/README.md:1]: Ok added it

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: Removed it

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: changed it

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: I added an error block at the end, that calls `psa_mac_abort` the operation

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: Also added a goto error block

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/main.c]: Also added error block here

**Comment by athoelke** [comment]
If we are going to be concerned with ensuring that the error clean-up code is correct in this example, we should probably also ensure that the example destroys the temporary keys that are created on each call to `sp800_108_counter_hmac_kdf()` and `sp800_108_counter_cmac_kdf()` - on both the success and error paths?

File: examples/crypto/SP800-108_counter_KDF/README.md
Comment (bheeb) [inline examples/crypto/SP800-108_counter_KDF/README.md:38]: This is not correct. The 32 byte HMAC keys from main.c must be used instead.

Comment (mswarowsky) [reply examples/crypto/SP800-108_counter_KDF/README.md:39]: Oh yeah, thanks, die print out was not updated, fixed it

**Comment by mswarowsky** [comment]
Rebased again, is there anything more I can do to get this moving forward?

**Comment by athoelke** [comment]
> Rebased again, is there anything more I can do to get this moving forward?

Hi @mswarowsky - as the example is based on using the mbedtls implementation, I was hoping to have one of the mbedtls team review the example as well.

Let me check if that is likely to happen soon - if not, then I won't let that block merging the PR.

**Comment by gilles-peskine-arm** [comment]
This is still on my list, I'll try to review this week.

File: examples/crypto/SP800-108_counter_KDF/README.md
Comment (gilles-peskine-arm) [inline examples/crypto/SP800-108_counter_KDF/README.md:123]: Info: I wrote a [Python implementation](https://github.com/Mbed-TLS/mbedtls/pull/8698) and got the same data.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I've reviewed the code and the documentation. I won't review the build script because I don't know CMake.

I reviewed to the standards of Mbed TLS except for the code style. We may not have the same goals in the PSA example repository, so I don't necessarily expect all my comments to be addressed. I would prefer to at least resolve the comments marked [MAJOR], but Andrew has the final say.

**Review by mswarowsky — COMMENTED** [review]
Addressed the feedback from @gilles-peskine-arm

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:15]: Spelling of OUTPUT:

```suggestion
#define OUTPUT_LENGTH    (42)
```

File: examples/crypto/SP800-108_counter_KDF/CMakeLists.txt
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/CMakeLists.txt:2]: The copyright date now includes 2024:

```suggestion
# SPDX-FileCopyrightText: Copyright (c) 2023-2024 Nordic Semiconductor ASA
```

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:49]: This is now 32 characters longer than the other output lines, after the reformatting of the hex bytes output.

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (athoelke) [inline examples/crypto/SP800-108_counter_KDF/main.c:58]: This has the same issue as line 49

**Review by athoelke — COMMENTED** [review]
A couple of minor issues. The rework looks correct, Gilles might like to review the update.

As this PR introduces new files, squashing the PR makes incremental review more challenging. Please can you avoid squashing until the review is complete. I will rebase and squash the PR immediately prior to merging, as we don't need/want the edit history in these files..

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (gilles-peskine-arm) [inline examples/crypto/SP800-108_counter_KDF/main.c:17]: It's defined as 4 in the specification. The fact that it's `sizeof(uint32_t)` is a coincidence.

File: examples/crypto/SP800-108_counter_KDF/main.c
Comment (gilles-peskine-arm) [inline examples/crypto/SP800-108_counter_KDF/main.c:32]: Minor: here we have a “null terminator” (in a “null-terminated string”), where the word “null” is an ordinary English word (an adjective, meaning “having the value zero”). So “null” should be spelled in lowercase. “`NULL`” is the usual way of writing a _null pointer_ in C code.

**Review by gilles-peskine-arm — APPROVED** [review]
Thanks for addressing my comments. Looks good to me apart from some very minor readability issues.

**Review by mswarowsky — COMMENTED** [review]
Addressed the new comments, with a fixup commit

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Pull request merged**
